***
# Introduction
**Scripts used to process and analyze Liver short-read RNA-seq data (obtained from Yang et al) and cerebellum RNA-seq data**

## Packages
### Linux packages
```{bash}
fastqc/0.11.9
java/12.0.1
trimmomatic/0.39
STAR/2.7.5b
picard/2.20.8
samtools/1.10
deeptools/3.5.0
```

### R packages
```{r}
library(Rsubread) #/2.6.4
library(gtools) #/3.9.2
library(DESeq2) #/1.30.1
library(tibble) #/3.1.7
library(dplyr) #/1.0.9
```

# Data Pre-processing
```{bash}
##== linux command ==##
path="/path"
#text file with the names of all the RNA-seq samples
#RNAseq_names.txt
```

# Adapter trimming
```{bash}
mkdir -p ${path}/fastq/adapt_trim

for sample_names in $(cat RNAseq_names.txt)
do
#run1
  java -jar $TRIMMOJAR PE -threads 24 -phred33 ${path}/fastq/run1/${sample_names}_R1_001.fastq.gz \
  ${path}/fastq/run1/${sample_names}_R2_001.fastq.gz ${path}/fastq/adapt_trim/${sample_names}_trimmed_run1_R1_paired.fq.gz \
  ${path}/fastq/adapt_trim/${sample_names}_trimmed_run1_R1_unpaired.fq.gz ${path}/fastq/adapt_trim/${sample_names}_trimmed_run1_R2_paired.fq.gz \
  ${path}/fastq/adapt_trim/${sample_names}_trimmed_run1_R2_unpaired.fq.gz \
  ILLUMINACLIP:${path}/TruSeq_and_nextera_adapters.fa:3:50:10 LEADING:10 TRAILING:10 SLIDINGWINDOW:4:20 MAXINFO:50:0.8 MINLEN:25
#run2
  java -jar $TRIMMOJAR PE -threads 24 -phred33 ${path}/fastq/run2/${sample_names}_R1_001.fastq.gz \
  ${path}/fastq/run2/${sample_names}_R2_001.fastq.gz ${path}/fastq/adapt_trim/${sample_names}_trimmed_run2_R1_paired.fq.gz \
  ${path}/fastq/adapt_trim/${sample_names}_trimmed_run2_R1_unpaired.fq.gz ${path}/fastq/adapt_trim/${sample_names}_trimmed_run2_R2_paired.fq.gz \
  ${path}/fastq/adapt_trim/${sample_names}_trimmed_run2_R2_unpaired.fq.gz \
  ILLUMINACLIP:${path}/TruSeq_and_nextera_adapters.fa:3:50:10 LEADING:10 TRAILING:10 SLIDINGWINDOW:4:20 MAXINFO:50:0.8 MINLEN:25
#remove unpaired fastqs
  rm ${path}/fastq/adapt_trim/*_unpaired.fq.gz
done
```

# Alignment using STAR
```{bash}
#calculate indexes using genomeGenerate (mm10)
#using genome fasta obtained from Gencode (GRCm38.p6)
mkdir ${path}/STAR_genome_mm10_ercc
chmod 777 ${path}/STAR_genome_mm10_ercc
mkdir ${path}/STAR_genome_mm10_ercc/star_genome  
chmod 777 ${path}/STAR_genome_mm10_ercc/star_genome
STAR --runMode genomeGenerate --genomeDir ${path}/STAR_genome_mm10_ercc/star_genome \
--genomeFastaFiles ${path}/mm10_ERCC_ref/GRCm38.primary_assembly.genome.fa --runThreadN 24

mkdir ${path}/bam
mkdir ${path}/bam/logs
chmod -R 777 ${path}/bam/logs

#Run1
for sample_names in $(cat RNAseq_names.txt)
do
   STAR --genomeDir ${path}/STAR_genome_mm10_ercc/star_genome \
   --readFilesIn ${path}/fastq/adapt_trim/${sample_names}_trimmed_run1_R1_paired.fq.gz \
   ${path}/fastq/adapt_trim/${sample_names}_trimmed_run1_R2_paired.fq.gz \
   --readFilesCommand zcat --outTmpDir ${path}/bam/logs/${sample_names}_STAR --outSAMunmapped Within \
   --outFilterType BySJout --outFilterMultimapNmax 20 --outFilterMismatchNmax 999 \
   --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 \
   --alignMatesGapMax 1000000 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --sjdbScore 1 --runThreadN 24\
   --outSAMtype BAM SortedByCoordinate \
   --quantMode TranscriptomeSAM --sjdbGTFfile ${path}/mm10_ERCC_ref/gencode_vM23_noM_annotation.gtf --sjdbGTFfeatureExon exon \
   --outFileNamePrefix ${path}/bam/${sample_names}_run1_
done

#Run2
for sample_names in $(cat RNAseq_names.txt)
do
   STAR --genomeDir ${path}/STAR_genome_mm10_ercc/star_genome \
   --readFilesIn ${path}/fastq/adapt_trim/${sample_names}_trimmed_run2_R1_paired.fq.gz \
   ${path}/fastq/adapt_trim/${sample_names}_trimmed_run2_R2_paired.fq.gz \
   --readFilesCommand zcat --outTmpDir ${path}/bam/logs/${sample_names}_STAR --outSAMunmapped Within \
   --outFilterType BySJout --outFilterMultimapNmax 20 --outFilterMismatchNmax 999 \
   --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 \
   --alignMatesGapMax 1000000 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --sjdbScore 1 --runThreadN 24\
   --outSAMtype BAM SortedByCoordinate \
   --quantMode TranscriptomeSAM --sjdbGTFfile ${path}/mm10_ERCC_ref/gencode_vM23_noM_annotation.gtf --sjdbGTFfeatureExon exon \
   --outFileNamePrefix ${path}/bam/${sample_names}_run2_
done
```

# Removing duplicates using Picard
#Obtained refFlat file from https://hgdownload.soe.ucsc.edu/goldenPath/mm10/database/refFlat.txt.gz
```{bash}
mkdir ${path}/bam/metric

for sample_names in $(cat RNAseq_names.txt)
do
   java -jar /usr/local/apps/picard/2.20.8/picard.jar  CollectRnaSeqMetrics \
   REF_FLAT=${path}/mm10_refFlat.txt \
   INPUT=${path}/bam/${sample_names}_run1_Aligned.sortedByCoord.out.bam \
   OUTPUT=${path}/bam/metric/${sample_names}_run1_RnaSeqMetrics.txt \
   STRAND=NONE
   
   java -jar /usr/local/apps/picard/2.20.8/picard.jar  CollectRnaSeqMetrics \
   REF_FLAT=${path}/mm10_refFlat.txt \
   INPUT=${path}/bam/${sample_names}_run2_Aligned.sortedByCoord.out.bam \
   OUTPUT=${path}/bam/metric/${sample_names}_run2_RnaSeqMetrics.txt \
   STRAND=NONE 

#remove duplicates
   java -jar /usr/local/apps/picard/2.20.8/picard.jar  MarkDuplicates \
   INPUT=${path}/bam/${sample_names}_run1_Aligned.sortedByCoord.out.bam \
   OUTPUT=${path}/bam/${sample_names}_run1_Aligned.out.MKDUP.bam \
   METRICS_FILE=${path}/bam/metric/${sample_names}_run1_RnaSeqMetrics.txt ASSUME_SORTED=true \
   MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=1000    REMOVE_DUPLICATES=true \
   REMOVE_SEQUENCING_DUPLICATES=true
   
   java -jar /usr/local/apps/picard/2.20.8/picard.jar  MarkDuplicates \
   INPUT=${path}/bam/${sample_names}_run2_Aligned.sortedByCoord.out.bam \
   OUTPUT=${path}/bam/${sample_names}_run2_Aligned.out.MKDUP.bam  \
   METRICS_FILE=${path}/bam/metric/${sample_names}_run2_RnaSeqMetrics.txt ASSUME_SORTED=true \
   MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=1000    REMOVE_DUPLICATES=true \
   REMOVE_SEQUENCING_DUPLICATES=true 
done
```

```{bash}
# Filter bam files
for sample_names in $(cat RNAseq_names.txt)
do
#run1
  samtools view -@ 12 -b -f 2 -q 10 ${path}/bam/${sample_names}_run1_Aligned.out.MKDUP.bam | \
  samtools view -@ 12 -b  >| ${path}/bam/${sample_names}_run1_filtered.bam
  
#run2
  samtools view -@ 12 -b -f 2 -q 10 ${path}/bam/${sample_names}_run2_Aligned.out.MKDUP.bam | \
  samtools view -@ 12 -b  >| ${path}/bam/${sample_names}_run2_filtered.bam
done


#Index bam files
for sample_names in $(cat RNAseq_names.txt)
do
#run1
  samtools index ${path}/bam/${sample_names}_run1_filtered.bam ${path}/bam/${sample_names}_run1_filtered_index.bam.bai
#run2
  samtools index ${path}/bam/${sample_names}_run2_filtered.bam ${path}/bam/${sample_names}_run2_filtered_index.bam.bai
done

#merge bam
mkdir -p ${path}/merged_bam

#Merge run 1 and run2 of RNA-seq using samtools
for sample_names in $(cat RNAseq_names.txt)
do
samtools merge ${path}/merged_bam/${sample_names}_merged.bam ${path}/bam/${sample_names}_run?_filtered.bam
done
```

# Generating bigwig files
```{bash}
#Index bam files
INPUT=${path}/bam/*_filtered.bam

for i in $INPUT
do
samtools index $i
done

#Generating bigwig files
mkdir -p ${path}/bigwig

for sample_names in $(cat $RNAseq_names.txt)
do
  bamCoverage -b ${path}/bam/${sample_names}_filtered.bam  -o ${path}/bigwig/${sample_names}_filtered.bw -of bigwig \
  --effectiveGenomeSize 1870000000 --extendReads 200 --normalizeUsing RPKM -p $SLURM_CPUS_PER_TASK
done
```

# Obtaining gene counts
```{r}
RNAseq_bam <-list.files("/path/bam", pattern="*_filtered.bam$", full.name=T)
RNAseq_bam <- mixedsort(RNAseq_bam)

#Using mm10 gencode comprehensive GTF (release M23, GRCm38.p6)
features <- featureCounts(files=RNAseq_bam,
        annot.ext="/path/gencode_vM23_noM_annotation.gtf",
        isGTFAnnotationFile = TRUE,
        GTF.featureType = "exon",
        GTF.attrType = "gene_name",
        GTF.attrType.extra = "gene_type",
        nthreads=12,
        isPairedEnd = T,checkFragLength = T,requireBothEndsMapped = T)

counts <- features$counts

sample_filename <- colnames(counts)
sample_name <- gsub("_filtered.bam","",sample_filename)
sample_name <- gsub("\\.","_",sample_name)
          
colnames(counts) <- sample_name
#Filtered for Liver RNA-seq before timepoint (but not 240h to have all genes matching before timepoint)
counts <- counts[rowSums(counts) > 1, ]

counts_DF <- as.data.frame(counts)
counts_DF <- tibble::rownames_to_column(counts_DF, "GeneID")

#only protein-coding genes
gene_type <- features$annotation[,c("GeneID", "gene_type")]
gene_type_proC <- gene_type[grepl("protein_coding", gene_type$gene_type),]   
counts_proC <- counts_DF %>% semi_join(gene_type_proC, by = "GeneID") # keep rows with matching ID

write.xlsx(counts_proC, "/path/RNAseq_gene_counttable.xlsx")
```

# Differential gene expression using DESeq2 (performed according to vignette by Michael I. Love, Simon Anders, and Wolfgang Huber)
#https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
```{r}
#Moving genes names to row
row.names(counts_proC) <- NULL
rownames(counts_proC)=counts_proC$GeneID
counts_proC=counts_proC[,-1]
head(counts_proC)

group <- c("old","old","old","old","young","young","young","young")
#Column 
Col <- as.data.frame(cbind(colnames(RNAseq),group))

#DESeq design
Col$group <- as.factor(Col$group)
Col$group <- relevel(Col$group, ref = "young")
dds <- DESeqDataSetFromMatrix(countData = RNAseq,
                              colData = Col,
                              design = ~group)
dds <- DESeq(dds)
resultsNames(dds)

#Differential analysis results
res_DF <- as.data.frame(res)
res_DF <- tibble::rownames_to_column(res_DF, "geneID")
write.xlsx(res_DF, "/path/DESeq2_DiffAnalysis.xlsx")
```

# within sample gene expression ranking for young and old
```{r}
#DESeq2 normalized counts
normalized_counts <- counts(dds, normalized=TRUE)
normalized_counts <- as.data.frame(normalized_counts)

#average young and old norm counts 
normalized_counts$Young_avg <- rowMeans(normalized_counts[ , c(5:8)], na.rm=TRUE)
normalized_counts$Old_avg <- rowMeans(normalized_counts[ , c(1:4)], na.rm=TRUE)

#Ranking genes for young and old by average
normalized_counts <- tibble::rownames_to_column(normalized_counts, "geneID")

#Young
RNAseq_quantiles <- normalized_counts %>%
  mutate(Y_quantile = ntile(Young_avg, 3))

#Old
RNAseq_quantiles <- RNAseq_quantiles %>%
  mutate(O_quantile = ntile(Old_avg, 3))
  
Young_low <- subset(RNAseq_quantiles, Y_quantile == "1")
Young_intermediate <- subset(RNAseq_quantiles, Y_quantile == "2")
Young_high <- subset(RNAseq_quantiles, Y_quantile == "3")

Old_low <- subset(RNAseq_quantiles, O_quantile == "1")
Old_intermediate <- subset(RNAseq_quantiles, O_quantile == "2")
Old_high <- subset(RNAseq_quantiles, O_quantile == "3")
```

# Ranking/categorizing DESeq2 results by fold change
```{r}
RNAseq_LFC <- res_DF %>%
  mutate(LFC_quantile = ntile(log2FoldChange, 3))

#subset
RNA_LFC_1 <- subset(RNAseq_LFC, LFC_quantile == "1")
RNA_LFC_3 <- subset(RNAseq_LFC, LFC_quantile == "3")

write.table(RNA_LFC_1$geneID, "/path/RNA_LFC_1.txt", row.names = FALSE, quote = FALSE, col.names = FALSE)
write.table(RNA_LFC_2$geneID, "/path/RNA_LFC_2.txt", row.names = FALSE, quote = FALSE, col.names = FALSE)
write.table(RNA_LFC_3$geneID, "/path/RNA_LFC_3.txt", row.names = FALSE, quote = FALSE, col.names = FALSE)
```

# Ranking/categorizing DESeq2 results by absolute fold change
```{r}
res_DF$foldchange_abs <- abs(res_DF$log2FoldChange)

RNAseq_absLFC <- res_DF %>%
  mutate(LFC_quantile_abs = ntile(foldchange_abs, 3))
  
#subset
RNA_absLFC_1 <- subset(RNAseq_absLFC, LFC_quantile_abs == "1")
RNA_absLFC_3 <- subset(RNAseq_absLFC, LFC_quantile_abs == "3")

write.table(RNA_absLFC_1$geneID, "/path/RNA_absLFC_1.txt", row.names = FALSE, quote = FALSE, col.names = FALSE)
write.table(RNA_absLFC_3$geneID, "/path/RNA_absLFC_3.txt", row.names = FALSE, quote = FALSE, col.names = FALSE)
```
